feat(ios): use/support/build xcframeworks

sgtcoolguy opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-28116

**Description:**
This is required for #11955 

These are the changes to move to using xcframeworks for APS, tiverify, TitaniumKit and for native modules.

This includes changes to the libraries:

- appcelerator/aps_sdk#347
- appcelerator/APSHTTPClient#54
- appcelerator/titanium_verify#33

Xcode minimum is bumped to 11 as that is required for xcframework support.

This PR does build/include the built maccatalyst binaries in the xcframeworks. It *does not* enable macOS as a target platform. #11955 will be used for that.

**Note** that to avoid breaking existing native modules we will selectively exclude arm64 from simulator builds of apps if we detect a native module that is not an xcframework - or is built with Xcode 11 and is missing arm64 simulator binaries.

If the developer is running an arm64 Mac build of Node (there are none yet officially!), and a native module is assumed to not have arm64 sim support, we'll error out with a message about it. (the "fix" there is for us/them to rebuild the native modules using Xcode 12 w/ xcframework support)

**TODO**
 - [x] Address @vijaysingh-axway review feedback around use of Info.plist from xcframework to determine arch/folder name support.
 - [x] make the arm64 sim exclusion smarter by handling if an xcframework has no arm64 sim support.

vijaysingh-axway committed:
build(ios): update APS libraries, tiverify to xcframework

- https://github.com/appcelerator/aps_sdk/pull/347
- https://github.com/appcelerator/APSHTTPClient/pull/54
- https://github.com/appcelerator/titanium_verify/pull/33

sgtcoolguy committed:
refactor(ios): guard APIs for mac catalyst target

sgtcoolguy committed:
chore(ios)!: set minimum xcode to 11.x

Fixes TIMOB-28072

sgtcoolguy committed:
fix(ios): exclude arm64 arch for sim target if native modules aren't xcframeworks

Modules built before xcframework support do not have an arm64 sim arch/target
To avoid having to break all modules, we'll try and exclude arm64 from sim builds
This will eventually break if app is built on arm64 apple silicon and a new
xcframework based module hasn't been shipped.

sgtcoolguy committed:
feat(ios): build modules as xcframeworks (w/ macos support)

- preserve symlinks in module xcframeworks/zips

sgtcoolguy committed:
feat(ios): use/support/build xcframeworks

build commented:
<!--
  1 failure:  :microscope: Ther...
  4 warning:  iphone/cli/hooks/..., iphone/cli/hooks/..., iphone/cli/hooks/..., iphone/cli/hooks/...
  3 messages
  
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/sgtcoolguy/titanium_mobile/blob/xcframeworks/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  **iphone/cli/hooks/frameworks.js#L60** - iphone/cli/hooks/frameworks.js line 60 – Avoid calling back inside of a promise. (promise/no-callback-in-promise)
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **iphone/cli/hooks/frameworks.js#L60** - iphone/cli/hooks/frameworks.js line 60 – Expected catch() or return (promise/catch-or-return)
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **iphone/cli/hooks/frameworks.js#L79** - iphone/cli/hooks/frameworks.js line 79 – Avoid wrapping return values in Promise.resolve (promise/no-return-wrap)
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **iphone/cli/hooks/frameworks.js#L429** - iphone/cli/hooks/frameworks.js line 429 – Each then() should return a value or throw (promise/always-return)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-12026/1/artifact/dist/mobilesdk-9.2.0.v20200909114234-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :white_check_mark: All tests are passing
Nice one! All 7445 tests are passing.
(There are 711 skipped tests not included in that total)
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1c5f8472841c764f2c8491affc3f23e1d95ad3b0
</p>

sgtcoolguy commented:
@vijaysingh-axway This is the xcframeworks only changes. I addressed your initial feedback re: Xcode 11 built xcframeworks in a lazy/dumb way - checking for the Xcode 12+ dir name and falling back to the Xcode 11 folder name (rather than try and get fancy with parsing Info.plist).

If we see an xcframework built on Xcode 11 missing arm64, we'll exclude arm64 on app simulator builds just as we would for a non-xcframework based native module.

vijaysingh-axway commented on iphone/Classes/ContactsModule.h:
```diff
@@ -6,7 +6,7 @@
 #import <TitaniumKit/TiModule.h>
 
-#ifdef USE_TI_CONTACTS
+#if defined(USE_TI_CONTACTS) && (!TARGET_OS_MACCATALYST || IS_SDK_IOS_14)
```
I see that TARGET_OS_MACCATALYST is getting used. I guess it is not removed intentionally to save time. I think it is fine.

sgtcoolguy replied:
Contacts API is available in Mac Catalyst 14+. So I intentionally guarded here like this.

vijaysingh-axway commented on iphone/cli/commands/_buildModule.js:
```diff
@@ -460,156 +459,202 @@ iOSModuleBuilder.prototype.buildModule = function buildModule(next) {
+				const osVersionParts = macOsVersion.split('.').map(n => parseInt(n));
+				if (osVersionParts[0] > 10 || (osVersionParts[0] === 10 && osVersionParts[1] >= 15)) {
+					// 3. Create a build for the maccatalyst
+					xcodebuildHook(xcBuild, xcodeBuildArgumentsForTarget('macosx'), opts, 'xcode-macos', next);
```
building module for macOS is intentional?

sgtcoolguy replied:
Yes, I wanted to future proof. We do still plan on landing macOS if we can. But if we don't this would avoid having to rebuild the module multiple times.

vijaysingh-axway commented on iphone/iphone/Titanium.entitlements:
```diff
@@ -1,5 +1,24 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
-<dict/>
+<dict>
```
These should be removed as we are excluding mac target in this PR.

sgtcoolguy replied:
Yeah, this should probably be in the macOS support PR.

vijaysingh-axway commented on iphone/iphone/Titanium.xcodeproj/project.pbxproj:
```diff
@@ -2406,9 +2417,15 @@
 				SWIFT_VERSION = 5.0;
+				PROVISIONING_PROFILE_SPECIFIER = "";
+				"PROVISIONING_PROFILE_SPECIFIER[sdk=macosx*]" = "";
+				SUPPORTS_MACCATALYST = YES;
```
Is it left intentional? Or we should remove this?

sgtcoolguy replied:
Intentional. Wanted to guard for Mac catalyst so modules could build for catalyst and just be ready without having to make them rebuild multiple times.

vijaysingh-axway commented on iphone/package.json:
```diff
@@ -19,7 +19,7 @@
 	"vendorDependencies": {
-		"xcode": ">=9.0 <=12.x",
+		"xcode": ">=11.0 <=12.x",
 		"ios sdk": ">=10.0 <=14.x"
```
```suggestion
		"ios sdk": ">=13.0 <=14.x"
```

sgtcoolguy replied:
Yeah, I think the min iOS bump can easily be done separately and iOS a bigger issue.

vijaysingh-axway reviewed (commented):
@sgtcoolguy  Changes looks good. I have left few comments. Basically questions. Can you look in that? Thanks!

sgtcoolguy commented:
@ewanharris I've incorporated your feedback. Updated the line for the framework search path sniffing, and right before we create-xcframework we'll wipe the destination if it exists to avoid the "couldn't be copied" error.
